### PR TITLE
Milking

### PR DESCRIPTION
**Pull request by AlecWhite**
Remade https://github.com/CleverRaven/Cataclysm-DDA/issues/21042 with the suggestions by @Coolthulhu, and the myriad of other advises given by everyone. Again, thank you for that.

- Cows are milkable but need to be tamed with "cattle fodder" tool.
- Added cattle food recipe, item and action.
- Cows can only be milking 3 times per day with a day of cool down between milkings.
- Tamed cows don't follow the player, they are too dumb and simple minded to do such things.
- Lowered cow's aggression added PET_WONT_FOLLOW and MILKABLE flags to cows too.
- Added a check in monmove.cpp for tamed animals with the SIMPLEMINDED flag to not follow the player.
- Remade(or is it called refactored?) the pet_food code to use the PetFood enum, and changed all the if to switches, this way is easier to add future items to tame other animals.
- Made two new files, monexamine.h/.cpp to take care of milking interaction and any other interactions with pets or monsters that we make in the future.
- Added the milking code to handle_liquids function and its action.

I think I haven't forgotten to mention anything important.

And as always, I ask for your review and help with fixing any possible problems please.

### CODE REVIEW

File: data/json/items/tools.json [hunk 1]
```diff
@@ -8270,5 +8270,20 @@
     "ammo": "battery",
     "initial_charges": 100,
     "max_charges": 100
+  },
+  {
+    "id": "cattlefodder",
+    "type": "TOOL",
+    "name": "cattle fodder",
+    "name_plural": "cattle fodder",
+    "description": "What cattle eats. Mainly made of grass, silages or legumes, is perfect for rumiants",
```
Comment (Coolthulhu) [inline data/json/items/tools.json]: Two spaces after period

File: data/json/item_actions.json [hunk 2]
```diff
@@ -731,5 +731,10 @@
         "type" : "item_action",
         "id" : "ups_based_armor",
         "name" : "Activate/deactivate"
+    },
+    {
+        "type" : "item_action",
+        "id" : "CATTLEFODDER",
+        "name" : "Place"
```
Comment (Coolthulhu) [inline data/json/item_actions.json]: "Feed cattle" would be better - "Place" sounds like you're placing it like a trap.

File: data/json/recipes/recipe_food.json [hunk 3]
```diff
@@ -10056,5 +10056,20 @@
...
+    "subcategory": "CSC_FOOD_OTHER",
+    "skill_used": "cooking",
+    "difficulty": 2,
+    "skills_required": [ "survival", 1 ],
+    "time": 10000,
+    "autolearn": true,
+    "components": [
+      [ [ "oats", 1 ], [ "buckwheat", 1 ], [ "wheat", 1 ], [ "potato_raw", 1 ], [ "barley", 1 ], [ "raw_beans", 1 ], [ "cabbage", 1 ], [ "veggy_wild", 1 ], [ "veggy", 1 ] ],
+      [ [ "oats", 1 ], [ "buckwheat", 1 ], [ "wheat", 1 ], [ "potato_raw", 1 ], [ "barley", 1 ], [ "raw_beans", 1 ], [ "cabbage", 1 ], [ "veggy_wild", 1 ], [ "veggy", 1 ] ],
+      [ [ "oats", 1 ], [ "buckwheat", 1 ], [ "wheat", 1 ], [ "potato_raw", 1 ], [ "barley", 1 ], [ "raw_beans", 1 ], [ "cabbage", 1 ], [ "veggy_wild", 1 ], [ "veggy", 1 ] ]
```
Comment (Coolthulhu) [inline data/json/recipes/recipe_food.json]: This will not work - the crafting code doesn't understand multiple components of the same type.

File: src/activity_handlers.cpp [hunk 4]
```diff
@@ -759,7 +786,22 @@ void activity_handlers::fill_liquid_do_turn( player_activity *act_, player *p )
         case LST_INFINITE_MAP:
             // nothing, the liquid source is infinite
             break;
-        }
+        case LST_MONSTER:
+            if( source_mon->has_effect( effect_milked ) ) {
+                current_dur = source_mon->get_effect_dur( effect_milked );
+            } if( !source_mon->has_effect( effect_milked ) ) {
```
Comment (Coolthulhu) [inline src/activity_handlers.cpp]: Weird spacing

File: src/activity_handlers.cpp [hunk 5]
```diff
@@ -759,7 +786,22 @@ void activity_handlers::fill_liquid_do_turn( player_activity *act_, player *p )
...
+                source_mon->add_effect( effect_milked, HOURS( 6 ) );
+            } else if( ( HOURS( 24 ) - current_dur ) < HOURS( 24 ) ) {
+                 source_mon->add_effect( effect_milked, ( current_dur + HOURS( 6 ) ) );
```
Comment (Coolthulhu) [inline src/activity_handlers.cpp]: Those extra parentheses aren't needed

File: src/game.cpp [hunk 6]
```diff
@@ -7218,12 +7219,13 @@ bool pet_menu(monster *z)
         drop_all,
         give_items,
         pheromone,
+		milk,
```
Comment (Coolthulhu) [inline src/game.cpp]: Tabs

File: src/monexamine.cpp [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+#include "map.h"
+#include "string_input_popup.h"
+#include "messages.h"
+
+const efftype_id effect_milked( "milked" );
+
+void monexamine::milk_source( monster *source_mon )
+{
+    item milk( "milk", 0, 1 );
+    if( !source_mon->has_effect( effect_milked ) ||
```
Comment (Coolthulhu) [inline src/monexamine.cpp]: You shouldn't need this case: if the creature has no effect of a particular ID, it has 0 duration of said effect, meaning it's handled by case below.

File: src/monexamine.h [hunk 8]
```diff
@@ -0,0 +1,11 @@
+#pragma once
+#ifndef MONEXAMINE
+#define MONEXAMINE
+
+class monster;
+
+namespace monexamine
+{
+void milk_source( monster *source_mon );
```
Comment (Coolthulhu) [inline src/monexamine.h]: You don't check if the `source_mon` is ever `nullptr` and don't expect `nullptr` arguments here, so a reference would be better here.

File: src/mtype.h [hunk 9]
```diff
@@ -157,6 +157,8 @@ enum m_flag : int {
     MF_AVOID_DANGER_2,      // This monster will path around most dangers instead of through them.
     MF_PRIORITIZE_TARGETS,  // This monster will prioritize targets depending on their danger levels
     MF_NOT_HALLU,           // Monsters that will NOT appear when player's producing hallucinations
+	MF_MILKABLE,            // This monster is milkable.
+	MF_SIMPLEMINDED,		// This monster is simpleminded.
```
Comment (Coolthulhu) [inline src/mtype.h]: The name is rather nondescriptive. How about `MF_PET_WONT_FOLLOW`?
Also, tabs - here and in other files.

Comment (AlecWhite) [inline src/mtype.h]: I think I may have hunted down all the tabs,

File: src/activity_handlers.cpp (hunk 4 above)
Comment (AlecWhite) [inline src/activity_handlers.cpp]: That spacing is the one the Astyling do. Unless I'm missing something.

File: src/iuse.cpp [hunk 10]
```diff
@@ -1901,63 +1902,84 @@ int iuse::mycus(player *p, item *it, bool t, const tripoint &pos)
...
+                if( mon.type->id == mon_dog_thing ) {
+                    p->deal_damage( &mon, bp_hand_r, damage_instance( DT_CUT, rng( 1, 10 ) ) );
+                    add_msg( m_bad, _( "You want to feed it the dog food, but it bites your fingers!" ) );
+                    if( one_in( 5 ) ) {
+                        add_msg(_( "Apparently it's more interested in your flesh than the dog food in your hand!" ) );
+                    }
+
+                } else if( npc *const person_ = g->critter_at<npc>( dirp ) ) {
+                    npc &person = *person_;
+                    query_yn( _( "Are you sure you want to feed a person the dog food?" ) );
```
Comment (BrettDong) [inline src/iuse.cpp]: This query is left useless?

File: src/activity_handlers.cpp [hunk 11]
```diff
@@ -696,6 +708,18 @@ void activity_handlers::fill_liquid_do_turn( player_activity *act_, player *p )
             on_ground = source_stack.begin();
             std::advance( on_ground, act.values.at( 1 ) );
             liquid = *on_ground;
+			break;
+        case LST_MONSTER:
+            Creature *c = g->critter_at( source_pos );
+            source_mon = dynamic_cast<monster *>( c );
+            if( source_mon == nullptr ) {
+                debugmsg( "could not find source creature for liquid transfer" );
```
Comment (BevapDin) [inline src/activity_handlers.cpp]: This should throw like in the other cases above. Currently it just continues with a null pointer, which is dereferenced later (for the `has_effect` check).

File: src/activity_handlers.cpp (hunk 4 above)
Comment (BevapDin) [inline src/activity_handlers.cpp]: That formating is very misleading. It *looks* like `} else if {`, but it is actually an unrelated `if`.
The `if` belong on the next line.

File: src/game.cpp [hunk 12]
```diff
@@ -7218,12 +7219,13 @@ bool pet_menu(monster *z)
...
         give_items,
         pheromone,
+        milk,
         rope
     };
 
     uimenu amenu;
     amenu.return_invalid = true;
-    std::string pet_name = _("dog");
+    std::string pet_name = _(z->get_name().c_str());
```
Comment (BevapDin) [inline src/game.cpp]: `monster::get_name` returns a translated string, no need to translate it again (get rid of the `_()` and the `.c_str()`).

File: src/monmove.cpp [hunk 13]
```diff
@@ -209,8 +209,8 @@ void monster::plan( const mfactions &factions )
     bool swarms = has_flag( MF_SWARMS );
     auto mood = attitude();
 
-    // If we can see the player, move toward them or flee.
-    if( friendly == 0 && sees( g->u ) ) {
+    // If we can see the player, move toward them or flee, simpleminded animals are too dumb to follow the player.
+    if( friendly == 0 && sees( g->u ) &&  !has_flag( MF_PET_WONT_FOLLOW ) ) {
```
Comment (BevapDin) [inline src/monmove.cpp]: There should only be one, not two spaces in front of the `!has_flag`.

File: src/monexamine.cpp [hunk 14]
```diff
@@ -0,0 +1,24 @@
+#include "monexamine.h"
+#include "monster.h"
+#include "string_input_popup.h"
+#include "monster.h"
+#include "calendar.h"
+#include "game.h"
+#include "map.h"
+#include "string_input_popup.h"
```
Comment (BevapDin) [inline src/monexamine.cpp]: Why is that needed (*again*)?

File: src/monexamine.cpp [hunk 15]
```diff
@@ -0,0 +1,24 @@
+#include "monexamine.h"
+#include "monster.h"
+#include "string_input_popup.h"
+#include "monster.h"
+#include "calendar.h"
+#include "game.h"
+#include "map.h"
```
Comment (BevapDin) [inline src/monexamine.cpp]: Why is that needed?

File: src/monexamine.cpp [hunk 16]
```diff
@@ -0,0 +1,24 @@
+#include "monexamine.h"
+#include "monster.h"
+#include "string_input_popup.h"
```
Comment (BevapDin) [inline src/monexamine.cpp]: Why is that needed?

File: src/monexamine.cpp [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+#include "monster.h"
```
Comment (BevapDin) [inline src/monexamine.cpp]: This is already included.

File: src/iuse.h [hunk 18]
```diff
@@ -68,7 +68,7 @@ class iuse
     int mycus               ( player*, item*, bool, const tripoint& );
     int dogfood             ( player*, item*, bool, const tripoint& );
     int catfood             ( player*, item*, bool, const tripoint& );
-
+	int feedcattle         ( player*, item*, bool, const tripoint& );
```
Comment (BevapDin) [inline src/iuse.h]: There is a tab, it should be 4 spaces.

File: src/iuse.cpp [hunk 19]
```diff
@@ -67,6 +67,7 @@ const mtype_id mon_shadow( "mon_shadow" );
 const mtype_id mon_spore( "mon_spore" );
 const mtype_id mon_vortex( "mon_vortex" );
 const mtype_id mon_wasp( "mon_wasp" );
+const mtype_id mon_cow("mon_cow");
```
Comment (BevapDin) [inline src/iuse.cpp]: Spaces are missing.

File: src/iuse.cpp [hunk 20]
```diff
@@ -1901,63 +1902,84 @@ int iuse::mycus(player *p, item *it, bool t, const tripoint &pos)
...
+        switch( animal_food_type ) {
+            case DOGFOOD:
+                if( mon.type->id == mon_dog_thing ) {
+                    p->deal_damage( &mon, bp_hand_r, damage_instance( DT_CUT, rng( 1, 10 ) ) );
+                    add_msg( m_bad, _( "You want to feed it the dog food, but it bites your fingers!" ) );
+                    if( one_in( 5 ) ) {
+                        add_msg(_( "Apparently it's more interested in your flesh than the dog food in your hand!" ) );
+                    }
+
+                } else if( npc *const person_ = g->critter_at<npc>( dirp ) ) {
```
Comment (BevapDin) [inline src/iuse.cpp]: Unrelated to the PR: This condition can never be `true` (assuming no bug): the above `if` checks for a monster at `dirp`, which implies there can't be any other creature, so no npc will be there.

File: src/item_factory.cpp [hunk 21]
```diff
@@ -590,6 +590,7 @@ void Item_factory::init()
     add_iuse( "WEATHER_TOOL", &iuse::weather_tool );
     add_iuse( "WEED_BROWNIE", &iuse::weed_brownie );
     add_iuse( "XANAX", &iuse::xanax );
+	add_iuse( "FEEDCATTLE", &iuse::feedcattle );
```
Comment (BevapDin) [inline src/item_factory.cpp]: This looks like a tab. Should be 4 spaces.

File: src/activity_handlers.cpp [hunk 22]
```diff
@@ -759,7 +786,22 @@ void activity_handlers::fill_liquid_do_turn( player_activity *act_, player *p )
         case LST_INFINITE_MAP:
             // nothing, the liquid source is infinite
             break;
-        }
+        case LST_MONSTER:
+            if( source_mon->has_effect( effect_milked ) ) {
+                current_dur = source_mon->get_effect_dur( effect_milked );
+            } if( !source_mon->has_effect( effect_milked ) ) {
+                source_mon->add_effect( effect_milked, HOURS( 6 ) );
+            } else if( HOURS( 24 ) - current_dur < HOURS( 24 ) ) {
```
Comment (BevapDin) [inline src/activity_handlers.cpp]: That's a funny looking condition. It's equivalent to `-current_dur < 0`, which is equivalent to `current_dur > 0`. Why does it need to be so complicated?

File: src/activity_handlers.cpp [hunk 23]
```diff
@@ -759,7 +786,22 @@ void activity_handlers::fill_liquid_do_turn( player_activity *act_, player *p )
...
+                 source_mon->add_effect( effect_milked, ( current_dur + HOURS( 6 ) ) );
```
Comment (BevapDin) [inline src/activity_handlers.cpp]: The parentheses around `current_dur + HOURS( 6 )` are not needed.

File: src/game.cpp [hunk 24]
```diff
@@ -7442,6 +7448,11 @@ bool pet_menu(monster *z)
         return true;
     }
 
+    if( milk == choice ){
```
Comment (BevapDin) [inline src/game.cpp]: Missing space between `)` and `{`.

File: src/game.cpp [hunk 25]
```diff
@@ -9651,14 +9662,16 @@ bool game::handle_liquid_from_container( item &container, int radius )
 
 extern void serialize_liquid_source( player_activity &act, const vehicle &veh, const itype_id &ftype );
 extern void serialize_liquid_source( player_activity &act, const tripoint &pos, const item &liquid );
+extern void serialize_liquid_source(player_activity &act, const monster &mon, const item &liquid);
```
Comment (BevapDin) [inline src/game.cpp]: Spaces missing. See declarations above.

File: src/activity_handlers.cpp (hunk 11 above)
Comment (AlecWhite) [inline src/activity_handlers.cpp]: I did a throw originally following  the example of the other cases, then Coolthulhu https://github.com/CleverRaven/Cataclysm-DDA/pull/21042#pullrequestreview-41364116 pointed out that it wasn't necessary. Should I change it to a throw?

Comment (Coolthulhu) [inline src/activity_handlers.cpp]: You can cancel it in some other way. `throw` was non-standard in our code, but if you don't want to track how does the code work, you can go back to throw.

File: src/activity_handlers.cpp (hunk 22 above)
Comment (AlecWhite) [inline src/activity_handlers.cpp]: >Why does it need to be so complicated?

Because I'm awfully bad at even the simplest of algorithm.

I think I may have fixed it now.

File: src/activity_handlers.cpp (hunk 11 above)
Comment (AlecWhite) [inline src/activity_handlers.cpp]: I don't fully(or even partially) understand how player_activity works, I thought that `act.set_to_null();` would simple stop the activity.

There was a `void finish( player *p );` but doesn't seems to have a definition and is private.

If you mind to please explain how player_activity and activity_id work I ill try to find a better way, otherwise I don't mind putting the throw back. I really don't have any idea what would be the most proper and correct way to handle this situation so any help is welcomed.

File: src/activity_handlers.cpp (hunk 22 above)
Comment (AlecWhite) [inline src/activity_handlers.cpp]: Nope, I was wrong, will work on it further.

File: src/iuse.cpp [hunk 26]
```diff
@@ -1901,63 +1902,108 @@ int iuse::mycus(player *p, item *it, bool t, const tripoint &pos)
...
-        } else if( is_dogfood && mon.type->id == mon_dog_thing ) {
-            p->deal_damage( &mon, bp_hand_r, damage_instance( DT_CUT, rng( 1, 10 ) ) );
-            p->add_msg_if_player( m_bad, _( "You want to feed it the dog food, but it bites your fingers!" ) );
-            if( one_in( 5 ) ) {
-                p->add_msg_if_player( _( "Apparently it's more interested in your flesh than the dog food in your hand!" ) );
-            }
+
+    if( npc *const person_ = g->critter_at<npc>( dirp ) ) {
+        npc &person = *person_;
+        query_yn( _( "Are you sure you want to feed a person the dog food?" ) );
```
Comment (BrettDong) [inline src/iuse.cpp]: This query is still hung in the air.

File: src/activity_handlers.cpp (hunk 11 above)
Comment (kevingranade) [inline src/activity_handlers.cpp]: Neither this code nor the rest of the code in this method should throw, nothing outside of json parsing should do so.

File: src/iuse.cpp (hunk 26 above)
Comment (AlecWhite) [inline src/iuse.cpp]: After a2ed466ab7dc8f95347073380019620d4f9b2075 when I compile and test it with NPCs, I get that query when applying dog food to NPCs.
 Unless I'm misunderstanding what you mean by "hung in the air".

File: src/activity_handlers.cpp (hunk 11 above)
Comment (AlecWhite) [inline src/activity_handlers.cpp]: Then I'll leave it as a debugmsg.

File: src/game.cpp [hunk 27]
```diff
@@ -7218,12 +7219,13 @@ bool pet_menu(monster *z)
...
         give_items,
         pheromone,
+        milk,
         rope
     };
 
     uimenu amenu;
     amenu.return_invalid = true;
-    std::string pet_name = _("dog");
+    std::string pet_name = (z->get_name());
```
Comment (Coolthulhu) [inline src/game.cpp]: Unneeded extra parentheses here

File: src/monexamine.cpp [hunk 28]
```diff
@@ -0,0 +1,27 @@
...
+            g->handle_liquid(milk, nullptr, 0, nullptr, nullptr, source_mon);
+            add_msg(_("You milk the %s."), source_mon->get_name().c_str());
+        }
+        else {
+            add_msg(_("The %s's udders run dry."), source_mon->get_name().c_str());
+        }
+    }
+}
+
+
```
Comment (Coolthulhu) [inline src/monexamine.cpp]: Spacing in this file is all wrong.
Would be good if you added it to `astyle_whitelist` and astyled it.

File: src/monexamine.h (hunk 8 above)
Comment (AlecWhite) [inline src/monexamine.h]: Made check and now the pointer is needed.

Comment (Coolthulhu) [inline src/monexamine.h]: You shouldn't make the pointer needed, it should be obvious from context that you can sometimes supply `null` there.
But you only use it on known monsters. In fact, the `pet_menu` will crash if you supply it with a `null` monster.

Turning pointers to references and back is easy and readable, keeping things as pointers everywhere only obfuscates the intent.

**Comment by BorkBorkGoesTheCode** [comment]
Will this appear as part of 0.D?

**Comment by AlecWhite** [comment]
Closed in favour of https://github.com/CleverRaven/Cataclysm-DDA/pull/21457
